proto/opencdc/v1/opencdc.proto

```diff
@@ -0,0 +1,83 @@
+syntax = "proto3";
+
+package opencdc.v1;
+
+import "google/protobuf/descriptor.proto";
+import "google/protobuf/struct.proto";
+
+// CreatedAt can contain the time when the record was created in the 3rd party
+// system. The expected format is a unix timestamp in nanoseconds.
+option (metadata_created_at) = "opencdc.createdAt";
+// ReadAt can contain the time when the record was read from the 3rd party
+// system. The expected format is a unix timestamp in nanoseconds.
+option (metadata_read_at) = "opencdc.readAt";
+
+// We are (ab)using custom file options to define constants.
+// See https://github.com/protocolbuffers/protobuf/issues/3520#issuecomment-323613839
+extend google.protobuf.FileOptions {
+  string metadata_created_at = 1000;
+  string metadata_read_at = 1001;
+}
+
+// Operation defines what triggered the creation of a record.
+enum Operation {
+  OPERATION_UNSPECIFIED = 0;
+  // Records with operation create contain data of a newly created entity.
+  OPERATION_CREATE = 1;
+  // Records with operation update contain data of an updated entity.
+  OPERATION_UPDATE = 2;
+  // Records with operation delete contain data of a deleted entity.
+  OPERATION_DELETE = 3;
+  // Records with operation snapshot contain data of a previously existing
+  // entity, fetched as part of a snapshot.
+  OPERATION_SNAPSHOT = 4;
+}
+
+// Record contains data about a single change event related to a single entity.
+message Record {
+  // Position uniquely identifies the record.
+  bytes position = 1;
+
+  // Operation defines what triggered the creation of a record. There are four
+  // possibilities: create, update, delete or snapshot. The first three
+  // operations are encountered during normal CDC operation, while "snapshot" is
+  // meant to represent records during an initial load. Depending on the
+  // operation, the record will contain either the state before the change,
+  // after the change, or both (see fields before and after).
+  Operation operation = 2;
+
+  // Metadata contains optional information related to the record. Although the
+  // map can contain arbitrary keys, the standard provides a set of standard
+  // metadata fields (see options prefixed with metadata_*).
+  map<string, string> metadata = 3;
```

Comment:
Maybe I asked about this before, but can we make the values a bit more "generic" (i.e. not limit to strings only), so we have less work around parsing let's say numerical values?

Reply:
I think allowing more than just strings can complicate the handling of metadata, especially if we would allow things like arrays or maps.

I see metadata in a similar way like headers in a HTTP request, the values there are also encoded as strings. This makes it simpler to write code that doesn't care about the type of the metadata (e.g. logging), while allowing the code that does care about the type encode/decode it as it sees fit.

This doesn't mean we shouldn't make it easier to work with the correct types in code. We should provide utilities for standard metadata fields that will make it easier to set/get those fields. For example `opencdc.createdAt` is supposed to be a UNIX timestamp, so we can provide functions that set/get `time.Time`. The user doesn't have to worry about how the field is stored in the metadata map. (NB: we would have to supply these utilities either way even if we allow arbitrary types, because you have to cast the value and know which type to use).

Reply:
I'm fine with utilities too. I believe that will help with commonly used non-string types.